### PR TITLE
Fix #4748: Optimize JS methods

### LINKED ISSUE #4748: Optimize JS methods
When working on Exports, it seems to me we have potential for inlining a lot of trivial export forwarders. This might have real size impact on export heavy codebases.

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
TODO (before or after this PR): Give JSPropertyDefs a hash.

**Comment by gzm0** [comment]
Diff on the test suite: https://gist.github.com/gzm0/e208b870c4342c9d7795ba15f1d7e786

**Comment by gzm0** [comment]
Note to self: JS ctors are optimized but the optimized version is not used.

**Comment by gzm0** [comment]
The CI failures are because the `thisType` calculated is wrong: For JS classes, it uses `ClassType(...)` instead of `AnyType`.

This was not a problem before, because JS classes have no other instance methods.

**Comment by gzm0** [comment]
Version that appears to be working.

Diff on the test suite: https://gist.github.com/gzm0/5bf6365459c3b09cca02d235bd9063fa

most notable (IMO): It seems we're able to fully remove unused outer pointer fields from JS classes.

**Comment by sjrd** [comment]
The diff sure looks nice! :)

**Comment by gzm0** [comment]
updated: don't use auto-application (in IncOptimizer)

**Comment by gzm0** [comment]
Rebased

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala [hunk 1]
```diff
@@ -703,6 +777,9 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
       }
     }
 
+    private val jsMethodContainer = new JSMethodContainer(className, this)
+    jsMethodContainer.updateWith(linkedClass)
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:781]: Putting the `updateWith` call here is not consistent with `MethodContainer`, for which we put it in the constructor of `MethodContainer`. This confused me for a moment. Could we reconcile them?

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala [hunk 2]
```diff
@@ -721,15 +798,21 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
     private var fieldsRead: Set[FieldName] = linkedClass.fieldsRead
     private var staticFieldsRead: Set[FieldName] = linkedClass.staticFieldsRead
 
-    /* Class properties that can be retrieved without tracking.
+    /** This type of instances of this interface.
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:801]:
```suggestion
    /** The type of instances of this interface.
```
?

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala [hunk 3]
```diff
@@ -1063,15 +1154,102 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
...
+      updateDef(linkedMethod)
+
+    protected def doProcess(): JSConstructorDef = {
+      val JSConstructorDef(flags, params, restParam, body) = originalDef
+
+      val thisType = owner.untrackedThisType(flags.namespace)
+
+      val (newParamsAndRest, newRawBody) = new Optimizer(this, this.toString()).optimize(
+          None, thisType, params ++ restParam.toList, owner.untrackedJSClassCaptures, AnyType,
+          Block(body.allStats)(NoPosition), isNoArgCtor = false)
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:1231]:
```suggestion
          Block(body.allStats)(body.pos), isNoArgCtor = false)
```

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 4]
```diff
@@ -4450,8 +4451,21 @@ private[optimizer] abstract class OptimizerCore(
   }
 
   private def transformMethodDefBody(optTarget: Option[MethodID], thisType: Type,
-      params: List[ParamDef], resultType: Type, body: Tree,
-      isNoArgCtor: Boolean): (List[ParamDef], Tree) = {
+      params: List[ParamDef], jsClassCaptures: List[ParamDef], resultType: Type,
+      body: Tree, isNoArgCtor: Boolean): (List[ParamDef], Tree) = {
+
+    val captureLocalDefs = for {
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:4457]:
```suggestion
    val jsClassCaptureLocalDefs = for {
```

File: linker/shared/src/main/scala/org/scalajs/linker/standard/LinkedClass.scala [hunk 5]
```diff
@@ -112,9 +112,15 @@ final class LinkedClass(
...
+      methods: List[Versioned[MethodDef]],
+      exportedMembers: List[Versioned[JSMethodPropDef]],
+      jsConstructorDef: Option[Versioned[JSConstructorDef]]
   ): LinkedClass = {
-    copy(methods = methods)
+    copy(
+        methods = methods,
+        exportedMembers = exportedMembers,
+        jsConstructorDef = jsConstructorDef
+    )
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/standard/LinkedClass.scala:123]: I guess this should now be in the new style, with a trailing comma:
```suggestion
    copy(
      methods = methods,
      exportedMembers = exportedMembers,
      jsConstructorDef = jsConstructorDef,
    )
```

**Review by sjrd — CHANGES_REQUESTED** [review]
Looks great! I only have minor comments.

File: linker/shared/src/main/scala/org/scalajs/linker/standard/LinkedClass.scala (hunk 5 above)
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/standard/LinkedClass.scala:123]: Scalastyle still fails on trailing commas :(

**Comment by gzm0** [comment]
I believe I have addressed all the comments.

File: linker/shared/src/main/scala/org/scalajs/linker/standard/LinkedClass.scala (hunk 5 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/standard/LinkedClass.scala]: Ouch.

**Review by sjrd — APPROVED** [review]
LGTM ! Thanks all lot for this.

**Pull request merged**